R/weight.R

```diff
@@ -68,6 +89,49 @@ setMethod("weightVariables", "VariableCatalog", function (x) {
     }
 })
 
+#' @rdname weightVariables
+#' @export
+`weightVariables<-` <- function(x, value) {
+    if (is.dataset(x)) {
+        varCat <- allVariables(x)
+    } else if (inherits(x,  "VariableCatalog")) {
+        varCat <- x
+    } else {
+        halt("Left hand side is not a Crunch Dataset or Variable Catalog.")
+    }
+    current <- crGET(shojiURL(varCat, "orders", "weights"))
+    new <- current
+    if (is.null(value)) {
+        new$graph <- NULL
+    } else {
+        if (is.variable(value) || length(value) == 1) {
+            value <- list(value)
```

Comment:
Compare to how `is.na<-` and `hiddenVariables<-` (especially) work. When you assign in a variable (or alias!), it adds the variable to the list. We should be consistent.

Reply:
So you want something like this?
```
weightVariables(ds) <- var1
weightVariables(ds) <- var2
weightVariables(ds)
> c("var1", "var2")
```
How would you remove variables from weightVariables under that pattern?

Reply:
Yeah. That's how `hiddenVariables<-` works, and that's how `is.na<-.Categories` works. I think I was basing that model off of how `is.na<-` works in base. 

I could see a few ways to do remove weight variables with that pattern. One would be `is.weight(var) <- FALSE`. I could invent others that involve subsetting catalogs and orders, but that one is probably enough, and pretty natural to think about if you're a user trying to do this. 

What do you think?

Reply:
I guess I think we should avoid impurity as much as possible. In this case if`weightVariables(ds) <- ds$var` either set or removed `ds$var` as a possible weight variable the behaviour of the function would change depending on the state of the dataset on the server which can lead to some tricky bugs, for instance: 
- A user sets the `weightVariable()` in  a script and the behaviour of that script changes depending on whether its been run an odd or even number of times
- Two users set `weightVariable(ds) <- ds$var` from separate sessions which would result in no weight variable being set

It also feels a little weird to have the assignment operator remove something. In most cases it seems like assigning either creates or overwrites a reference, and if you want to remove a reference you should assign NULL. 

The package doesn't seem to be totally consistent on this point. The example I was emulating was `weight(ds) <- ds$var` which doesn't  remove `ds$var` as the weighting variable. 

In terms of the `is.weight(ds$var) <- TRUE` pattern, it seems like the main user story for this function is a dataset administrator setting the possible weight variables that a viewer can access. So my intuition is that `weightVariables` is more of a property of datasets, then it is of the variables within the dataset so it makes more sense to have the function operate on the dataset and allow them to set several variables in one step. It also might cause some confusion over whether the variable was allowed to be set as a weight or was currently set as the weight. Do we have users who are asking for this identifier function?

What about changing it so that it accepts a character vector of aliases instead of a list of variables? That way they could append variables with `weightVariables(ds) <- c(weightVariables(ds), "var")` and remove a variable with `weightVariables(ds) <- setdiff(weightVariables(ds), "var")`

Reply:
I was not advocating that `weightVariables(ds) <- ds$var` would set and then doing `weightVariables(ds) <- ds$var` again would remove. Nothing in the package works like that. I was suggesting, as you stated it before, 

```
weightVariables(ds) <- var1
weightVariables(ds) <- var2
weightVariables(ds)
> c("var1", "var2")
```

Yes, the assignment function should accept aliases.